### PR TITLE
[SPARK-32677][SQL] Load function resource before create

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Change `CreateFunctionCommand` code that add class check before create function.

### Why are the changes needed?

We have different behavior between create permanent function and temporary function when function class is invaild. e.g.,
```
create function f as 'test.non.exists.udf';
-- Time taken: 0.104 seconds

create temporary function f as 'test.non.exists.udf'
-- Error in query: Can not load class 'test.non.exists.udf' when registering the function 'f', please make sure it is on the classpath;
```

And Hive also fails both of them.

### Does this PR introduce _any_ user-facing change?

Yes, user will get exception when create a invalid udf.

### How was this patch tested?

New test.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUDFDynamicLoadSuite.scala [hunk 1]
```diff
@@ -156,7 +156,9 @@ class HiveUDFDynamicLoadSuite extends QueryTest with SQLTestUtils with TestHiveS
 
           sql(s"CREATE FUNCTION ${udfInfo.funcName} AS '${udfInfo.className}' USING JAR '$jarUrl'")
 
-          assert(Thread.currentThread().getContextClassLoader eq sparkClassLoader)
```
Comment (ulysses-you) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUDFDynamicLoadSuite.scala:159]: We need load resource during create function.

Comment (ulysses-you) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUDFDynamicLoadSuite.scala:159]: This test was created in [#27025](https://github.com/apache/spark/pull/27025).

**Comment by ulysses-you** [comment]
It's ok to review.   cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala [hunk 2]
```diff
@@ -76,14 +76,15 @@ case class CreateFunctionCommand(
   override def run(sparkSession: SparkSession): Seq[Row] = {
     val catalog = sparkSession.sessionState.catalog
     val func = CatalogFunction(FunctionIdentifier(functionName, databaseName), className, resources)
+    catalog.loadFunctionResources(resources)
+    // We fail fast if function class is not exists.
+    catalog.requireFunctionClassExists(func)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala:81]: is this needed? shouldn't `loadFunctionResources` already does the work?

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala:81]: `loadFunctionResources` only check the resource is exists, e.g. jar / file. Not include function class.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala:81]: can we register the function when creating it? That should load the class.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala:81]: For a permanent function, does it better cache when query on it? Seems the most use case is as a Hive client.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -1369,6 +1369,20 @@ class SessionCatalog(
     functionRegistry.registerFunction(func, info, builder)
   }
 
+  /**
+   * Make sure function class is on the classpath.
+   */
+  def requireFunctionClassExists(funcDefinition: CatalogFunction): Unit = {
+    val className = funcDefinition.className
+    if (!Utils.classIsLoadable(className)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1377]: can we reuse the code with `registerFunction`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 4]
```diff
@@ -1359,16 +1359,23 @@ class SessionCatalog(
...
+      throw new AnalysisException(s"Can not load class '$className' when registering " +
+        s"the function '${funcDefinition.identifier}', please make sure it is on the classpath")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1375]: how about `funcDefinition.identifier.unquotedString`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala:81]: I think we only need to do this for non-temp functions.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala:81]: Yeah, I also think about it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 4 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: sounds good.

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: `FunctionIdentifier` already override `toString = unquotedString`, but it's fine to invoke `unquotedString` explicit.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 5]
```diff
@@ -1359,16 +1359,24 @@ class SessionCatalog(
...
   }
 
+  /**
+   * Make sure function class is on the classpath.
+   */
+  def requireFunctionClassExists(funcDefinition: CatalogFunction): Unit = {
+    val className = funcDefinition.className
+    if (!Utils.classIsLoadable(className)) {
+      throw new AnalysisException(s"Can not load class '$className' when registering " +
+        s"the function '${funcDefinition.identifier.unquotedString}', please make sure " +
```
[the remark this answers is not in the document]
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1375]: Not needed. 

This method is used by both temporary and permanent function. The temporary has no database name and we can't fill the database name. The permanent follows user created.

File: sql/core/src/test/resources/sql-tests/results/udf/udf-udaf.sql.out [hunk 6]
```diff
@@ -51,7 +52,7 @@ SELECT default.udaf1(udf(int_col1)) as udaf1, udf(default.udaf1(udf(int_col1)))
 struct<>
 -- !query output
 org.apache.spark.sql.AnalysisException
-Can not load class 'test.non.existent.udaf' when registering the function 'default.udaf1', please make sure it is on the classpath; line 1 pos 94
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/udf/udf-udaf.sql.out:54]: Here it was `'default.udaf1'` previously. But the new message is `'udaf1'`: https://github.com/apache/spark/pull/29502/files#diff-2d3d6b47e6044b4d44590c5a73b7cd8bR46

Do you know why?

Comment (ulysses-you) [reply sql/core/src/test/resources/sql-tests/results/udf/udf-udaf.sql.out:54]: For permanent, the old code path is:
```
create function // no class check
query on function -> lookup function -> fill database name if permanent -> register function -> check class
```
So the previously msg always contains database name.

Now the code path is:
```
create function -> check class
```

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/udf/udf-udaf.sql.out:54]: then shall we fill in the default database when creating permanent functions, to make the error message better?

Comment (ulysses-you) [reply sql/core/src/test/resources/sql-tests/results/udf/udf-udaf.sql.out:54]: Permanent function always need a database, it's ok to fill database name when creating.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thanks for merging !

**Comment by ulysses-you** [comment]
Never mind. 

That said, there are two Spark cluster(A, B) and one Hive metastore then create function with Spark A but using Spark B's local path. Seems strange but it actual exists.

**Pull request closed without merging**